# Exercise - RNN Classification

In this notebook, we will perform a classification task using RNNs (i.e., a sequence to value prediction). We have hourly power consumption of households for 12 hours. Based on this, we will determine whether the power grid is strained (1) or not (0). 

Therefore, use the columns from `Hour 0` to `Hour 11` to predict the `target` column in the `power.csv` data set.

Hint1: Use Tutorial 1 for help.

Hint2: Don't forget to adjust the number of neurons in the input layers correctly. Otherwise, you will run into errors.

In [5]:
# Common imports

import tensorflow as tf
from tensorflow import keras

import numpy as np
import pandas as pd

# to make this notebook's output stable across runs
np.random.seed(42)


# Read the Dataset

In [6]:
power = pd.read_csv('power.csv')

power.head()

Hour 0    Hour 1    Hour 2    Hour 3    Hour 4    Hour 5    Hour 6  \
0  2.550633  2.523400  2.582333  2.541667  2.475733  2.476233  2.455800   
1  1.596933  1.619567  2.473733  2.731133  2.431133  2.479667  1.690200   
2  0.534933  0.540467  0.575367  0.526500  0.521900  0.565333  1.426467   
3  1.085867  0.651233  0.634600  0.653000  0.646067  0.628400  0.611067   
4  0.456000  0.286300  0.310833  0.250933  0.277667  0.308633  0.610400   

     Hour 7    Hour 8    Hour 9   Hour 10   Hour 11  target  
0  2.447200  2.441733  3.146133  2.661733  2.576000       1  
1  1.332133  1.375167  1.050900  0.585900  2.651900       1  
2  0.602067  0.547433  0.525067  1.270300  0.393767       0  
3  0.612533  0.660100  0.606067  1.471867  0.834533       0  
4  1.563533  1.421867  3.324400  3.207567  1.425433       1

In [7]:
power.shape

(1417, 13)

# Split the Data



In [8]:
from sklearn.model_selection import train_test_split
 
train, test = train_test_split(power, test_size=0.3)


In [9]:
train.shape

(991, 13)

In [10]:
test.shape

(426, 13)

# Create Input and Target values

The first 12 columns (hourly data) will be input to predict the last column (i.e., target)

In [11]:
# The first 12 columns (from 0 to 11) are inputs

train_x = train.iloc[:,:12]

## Set the target

In [12]:
# The last column is TARGET

train_y = train.iloc[:,-1]

## Repeat for TEST

In [13]:
test.shape

(426, 13)

In [14]:
# The first 12 columns are inputs

test_x = test.iloc[:,:12]

In [15]:
# The last column is TARGET

test_y = test.iloc[:,-1]

# Build a normal (cross-sectional) NN

This model assumes that the data is NOT a time-series data set. It treats the data as cross-sectional and the columns being independent of each other.

In [16]:
model = keras.models.Sequential([
    
    keras.layers.Flatten(input_shape=[12, 1]),
    keras.layers.Dense(12, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
    
])

C:\Users\mural\anaconda3\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [17]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = tf.keras.optimizers.Adam(learning_rate=0.01)

# If multiclass, use "sparse_categorical_crossentropy" as the loss function
model.compile(loss="binary_crossentropy", optimizer=optimizer, metrics=['accuracy'])


history = model.fit(train_x, train_y, epochs=50,
                    validation_data=(test_x, test_y))

Epoch 1/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 8ms/step - accuracy: 0.5862 - loss: 0.7017 - val_accuracy: 0.6596 - val_loss: 0.6398
Epoch 2/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7209 - loss: 0.6022 - val_accuracy: 0.6714 - val_loss: 0.5825
Epoch 3/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7467 - loss: 0.5512 - val_accuracy: 0.6995 - val_loss: 0.5586
Epoch 4/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7557 - loss: 0.5206 - val_accuracy: 0.7089 - val_loss: 0.5431
Epoch 5/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7687 - loss: 0.4990 - val_accuracy: 0.7019 - val_loss: 0.5348
Epoch 6/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7579 - loss: 0.4853 - val_accuracy: 0.7042 - val_loss: 0.5310
Epoch 7/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7610 - loss: 0.4770 - val_accuracy: 0.7136 - val_loss: 0.5290
Epoch 8/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7576 - loss: 0.4714 - val_accuracy: 0.7089 - val_loss:

In [18]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=2)

scores

# In results, first is loss, second is accuracy

14/14 - 0s - 4ms/step - accuracy: 0.7254 - loss: 0.5341


[0.5340527892112732, 0.7253521084785461]

In [19]:
model.evaluate

<bound method TensorFlowTrainer.evaluate of <Sequential name=sequential, built=True>>

In [15]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.54
compile_metrics: 72.30%


In [20]:
scores[0]

0.5340527892112732

In [21]:
scores[1]

0.7253521084785461

# Build a simple RNN with one layer

In [37]:
n_steps = 12
n_inputs = 1


model = keras.models.Sequential([
    
    keras.layers.SimpleRNN(32, input_shape=[n_steps, n_inputs]), # 32 is no of neurons just select multiples of 16 can even be greater than 
    keras.layers.Dense(1, activation='sigmoid')
])

In [25]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = tf.keras.optimizers.Adam(learning_rate=0.01)

# If multiclass, use "sparse_categorical_crossentropy" as the loss function
model.compile(loss="binary_crossentropy", optimizer=optimizer, metrics=['accuracy'])


history = model.fit(train_x, train_y, epochs=50,
                    validation_data=(test_x, test_y), callbacks=callback)

Epoch 1/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - accuracy: 0.6837 - loss: 0.5884 - val_accuracy: 0.7324 - val_loss: 0.5205
Epoch 2/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7382 - loss: 0.4881 - val_accuracy: 0.7277 - val_loss: 0.5381
Epoch 3/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7331 - loss: 0.5081 - val_accuracy: 0.7324 - val_loss: 0.5238
Epoch 4/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7499 - loss: 0.4920 - val_accuracy: 0.7277 - val_loss: 0.5232
Epoch 5/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7531 - loss: 0.4938 - val_accuracy: 0.7394 - val_loss: 0.5150
Epoch 6/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7519 - loss: 0.4899 - val_accuracy: 0.7324 - val_loss: 0.5376
Epoch 7/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7503 - loss: 0.5009 - val_accuracy: 0.7606 - val_loss: 0.5249
Epoch 8/50
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7497 - loss: 0.5022 - val_accuracy: 0.7230 - val_loss

In [26]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.5664351582527161, 0.7323943376541138]

In [27]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

loss: 0.57
compile_metrics: 73.24%


In [28]:
# Predictions are probabilities.

predictions = model.predict(test_x)

14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


In [29]:
# Rounding the probabilities determines 1 or 0

np.round(predictions)

array([[0.],
       [1.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [1.],
       [0.],
       [1.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [1.],
       [1.],
       [0.],
       [1.],
       [0.],
       [1.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [1.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [1.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [0.],
       [0.],
       [0.],
       [1.],
       [0.],

In [30]:
from sklearn.metrics import confusion_matrix

confusion_matrix(test_y, np.round(predictions))

array([[178,  30],
       [ 84, 134]], dtype=int64)

# Build a simple RNN with two or more layers

In [51]:
n_steps = 12
n_inputs = 1


model = keras.models.Sequential([
    keras.layers.SimpleRNN(16, return_sequences=True, input_shape=[n_steps, n_inputs] ),
    keras.layers.SimpleRNN(16, return_sequences=True),
    keras.layers.SimpleRNN(16), 
    keras.layers.Dense(1, activation='sigmoid')
])

In [52]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Adam(learning_rate=0.01)

model.compile(loss="binary_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 4s 22ms/step - accuracy: 0.6583 - loss: 0.6141 - val_accuracy: 0.7371 - val_loss: 0.5392
Epoch 2/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7292 - loss: 0.5155 - val_accuracy: 0.7113 - val_loss: 0.5288
Epoch 3/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7663 - loss: 0.4790 - val_accuracy: 0.7113 - val_loss: 0.5330
Epoch 4/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7502 - loss: 0.4755 - val_accuracy: 0.7230 - val_loss: 0.5185
Epoch 5/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.7705 - loss: 0.4724 - val_accuracy: 0.7653 - val_loss: 0.5029
Epoch 6/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7683 - loss: 0.4757 - val_accuracy: 0.7488 - val_loss: 0.5181
Epoch 7/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7703 - loss: 0.4700 - val_accuracy: 0.7207 - val_loss: 0.5377
Epoch 8/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7690 - loss: 0.4823 - val_accuracy: 0.7019 - val_loss

In [53]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.534654974937439, 0.7183098793029785]

In [54]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.53
compile_metrics: 71.83%


# Build a LSTM with one layer

In [55]:
n_steps = 12
n_inputs = 1

model = keras.models.Sequential([
    
    keras.layers.LSTM(32, input_shape=[n_steps, n_inputs]),
    keras.layers.Dense(1, activation='sigmoid')
])

In [56]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Adam(learning_rate=0.01)

model.compile(loss="binary_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 2s 16ms/step - accuracy: 0.6019 - loss: 0.6498 - val_accuracy: 0.7136 - val_loss: 0.5440
Epoch 2/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7524 - loss: 0.4984 - val_accuracy: 0.7441 - val_loss: 0.5231
Epoch 3/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7505 - loss: 0.4783 - val_accuracy: 0.7347 - val_loss: 0.5183
Epoch 4/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7602 - loss: 0.4813 - val_accuracy: 0.7465 - val_loss: 0.5223
Epoch 5/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7660 - loss: 0.4827 - val_accuracy: 0.7465 - val_loss: 0.5180
Epoch 5: early stopping


In [57]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.5179795622825623, 0.7464788556098938]

In [58]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.52
compile_metrics: 74.65%


# Build a LSTM with two or more layers

In [60]:
n_steps = 12
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.LSTM(32, return_sequences=True, input_shape=[n_steps, n_inputs]),
    keras.layers.LSTM(32, return_sequences=True),
    keras.layers.LSTM(32), #---> last layer return sequence must be false
    keras.layers.Dense(1, activation='sigmoid')
])


In [61]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Adam(learning_rate=0.01)

model.compile(loss="binary_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 5s 27ms/step - accuracy: 0.5577 - loss: 0.6548 - val_accuracy: 0.7441 - val_loss: 0.5803
Epoch 2/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.7457 - loss: 0.5486 - val_accuracy: 0.7324 - val_loss: 0.5622
Epoch 3/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.7497 - loss: 0.5084 - val_accuracy: 0.7441 - val_loss: 0.5388
Epoch 4/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.7576 - loss: 0.4950 - val_accuracy: 0.7371 - val_loss: 0.5202
Epoch 5/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.7523 - loss: 0.4776 - val_accuracy: 0.7300 - val_loss: 0.5169
Epoch 5: early stopping


In [62]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.516947865486145, 0.7300469279289246]

In [63]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.52
compile_metrics: 73.00%


# Build a GRU with one layer

In [64]:
n_steps = 12
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.GRU(32, input_shape=[n_steps, n_inputs]),
    keras.layers.Dense(1, activation='sigmoid')
])

In [65]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Adam(learning_rate=0.01)

model.compile(loss="binary_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 3s 20ms/step - accuracy: 0.6379 - loss: 0.6114 - val_accuracy: 0.7418 - val_loss: 0.5250
Epoch 2/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7397 - loss: 0.4822 - val_accuracy: 0.7394 - val_loss: 0.5272
Epoch 3/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7606 - loss: 0.4960 - val_accuracy: 0.7441 - val_loss: 0.5160
Epoch 4/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7619 - loss: 0.4818 - val_accuracy: 0.7465 - val_loss: 0.5152
Epoch 5/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7690 - loss: 0.4730 - val_accuracy: 0.7512 - val_loss: 0.5172
Epoch 5: early stopping


In [66]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.5172337293624878, 0.751173734664917]

In [67]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.52
compile_metrics: 75.12%


# Build a GRU with two or more layers

In [68]:
n_steps = 12
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.GRU(32, return_sequences=True, input_shape=[n_steps, n_inputs]),
    keras.layers.GRU(32, return_sequences=True),
    keras.layers.GRU(32),
    keras.layers.Dense(1, activation='sigmoid')
])

In [69]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Adam(learning_rate=0.01)

model.compile(loss="binary_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 7s 34ms/step - accuracy: 0.6598 - loss: 0.5988 - val_accuracy: 0.7324 - val_loss: 0.5291
Epoch 2/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step - accuracy: 0.7244 - loss: 0.5036 - val_accuracy: 0.7183 - val_loss: 0.5239
Epoch 3/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 16ms/step - accuracy: 0.7376 - loss: 0.4836 - val_accuracy: 0.7441 - val_loss: 0.5320
Epoch 4/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 18ms/step - accuracy: 0.7491 - loss: 0.4867 - val_accuracy: 0.7418 - val_loss: 0.5412
Epoch 5/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step - accuracy: 0.7670 - loss: 0.4895 - val_accuracy: 0.7277 - val_loss: 0.5394
Epoch 5: early stopping


In [70]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.5394094586372375, 0.7276995182037354]

In [71]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.54
compile_metrics: 72.77%


## Dicussion: 

#### Cross Sectional NN:       72.30%
#### RNN With one Layer:      73.24%
#### RNN With More Layers:  71.83% ------> Lowest
#### LSTM With One Layer:    74.65%
#### LSTM With More Layers: 73.00%
#### GRU With One Layer:      75.12%- -----> Highest
#### GRU With More Layers:   72.77%

#### All models achieved moderate accuracy, ranging from 70% to 75%. This suggests the models can identify patterns in the data but may not be perfect at predicting the target variable.

#### Both LSTM with one layer (74.65%) and GRU with one layer (75.12%) outperformed models with more layers. This suggests that for this dataset, increasing model complexity might not necessarily improve performance.

#### The specific data that is based on for power requires handling long-term dependencies, which will benificial for predciting.